src/emulator/extensionsEmulator.ts

```diff
@@ -52,25 +53,47 @@ export class ExtensionsEmulator {
     }
     // TODO: If ref contains 'latest', we need to resolve that to a real version.
 
+    const ref = toExtensionVersionRef(instance.ref);
     const cacheDir =
       process.env.FIREBASE_EXTENSIONS_CACHE_PATH ||
       path.join(os.homedir(), ".cache", "firebase", "extensions");
-    const sourceCodePath = path.join(cacheDir, toExtensionVersionRef(instance.ref));
+    const sourceCodePath = path.join(cacheDir, ref);
 
-    // Check if something is at the cache location already. If so, assume its the extension source code!
-    // TODO(b/216376066): Add some better sanity checking that it is the extension source code & was successfully downloaded.
-    if (!fs.existsSync(sourceCodePath)) {
+    if (!this.hasValidSource({ path: sourceCodePath, extRef: ref })) {
       const extensionVersion = await planner.getExtensionVersion(instance);
-      await downloadExtensionVersion(
-        toExtensionVersionRef(instance.ref),
-        extensionVersion.sourceDownloadUri,
-        sourceCodePath
-      );
+      await downloadExtensionVersion(ref, extensionVersion.sourceDownloadUri, sourceCodePath);
       this.installAndBuildSourceCode(sourceCodePath);
     }
     return sourceCodePath;
   }
 
+  /**
+   * Returns if the source code at given path is valid.
+   *
+   * Checks against a list of required files or directories that need to be present.
+   */
+  private hasValidSource(args: { path: string; extRef: string }): boolean {
+    const requiredFiles = [
+      "./extension.yaml",
+      "./functions/package.json",
+      "./functions/node_modules",
```

Comment:
Technically, extensions can have their functions code somewhere other than the default 'functions', by setting 'sourceDirectory' on the function resource in extension.yaml. (see cs/piper///depot/google3/firebase/mods/go/worker/fetch_mod_source.go;l=451)

I'm not sure we need to implement support for this yet (IDK if any extensions even use it) but maybe add a todo?

Reply:
I see, didn't know that before, added a todo :)